opal/mca/btl/usnic/README.txt

```diff
@@ -335,3 +335,35 @@ libfabric abstractions:
 fi_fabric: corresponds to a VIC PF
 fi_domain: corresponds to a VIC VF
 fi_endpoint: resources inside the VIC VF (basically a QP)
+
+======================================
+
+MPI_THREAD_MULTIPLE support
+
+In order to make usnic btl thread-safe, the mutex locks are issued
+to protect the critical path. ie; libfabric routines, book keeping, etc.
+
+The said lock is btl_usnic_send_lock. It is a RECURSIVE lock, meaning that 
+the same thread can take the lock again even if it already has the lock to
+allow the callback function to post another segment right away if we know
+that the current segment is completed inline. (So we can call send in send
+without deadlocking)
+
+These two functions taking care of hotel checkin/checkout and we
+have to protect that part. So we take the mutex lock before we enter the
+function.
+
+- opal_btl_usnic_check_rts()  
+- opal_btl_usnic_handle_ack()  
+
+We also have to protect the call to libfabric routines
+
+- opal_btl_usnic_endpoint_send_segment()        (fi_send)
+- opal_btl_usnic_recv_call()			(fi_recvmsg)
+
+have to be protected as well.
+
```

Comment:
Don't forget to mention the cclient usage has to be protected (but not the cagent).